### PR TITLE
Text Detection: Add text detection metrics

### PR DESCRIPTION
**Pull request by the-star-sea**
- [x] add metrics(precision,recall,hmean,AP,avg time)
- [x] fix a small print error in  demo.py

### CODE REVIEW

**Comment by fengyuentau** [comment]
We dont put evaluation script along with models. Please take a look at https://github.com/opencv/opencv_zoo/tree/master/tools/eval.

**Comment by the-star-sea** [comment]
I wanna add some specific functions for detection task. I think we need a discussion and I have talked about it with zihao Mu.

File: models/text_detection_db/cal_rescall/rrc_evaluation_funcs.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    for line in lines:
+        line = line.replace("\r","").replace("\n","")
+        if(line != ""):
+            try:
+                validate_tl_line(line,LTRB,withTranscription,withConfidence,imWidth,imHeight)
+            except Exception as e:
+                raise Exception(("Line in sample not valid. Sample: %s Line: %s Error: %s" %(fileName,line,str(e))).encode('utf-8', 'replace'))
+    
+   
+
```
Comment (zihaomu) [inline models/text_detection_db/cal_rescall/rrc_evaluation_funcs.py:137]: Please keep two empty lines.

File: models/text_detection_db/cal_rescall/script.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+            detFile = subm[resFile]  # rrc_evaluation_funcs.decode_utf8(subm[resFile])
+
+            pointsList, confidencesList, _ = rrc_evaluation_funcs.get_tl_line_values_from_file_contents(detFile,
+                                                                                                        evaluationParams[
+                                                                                                            'CRLF'],
+                                                                                                        evaluationParams[
+                                                                                                            'LTRB'],
+                                                                                                        False,
+                                                                                                        evaluationParams[
+                                                                                                            'CONFIDENCES'])
```
Comment (zihaomu) [inline models/text_detection_db/cal_rescall/script.py:169]: wired format. Please use `\` for very long py codes.

File: models/text_detection_db/validation.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
+# This file is part of OpenCV Zoo project.
+# It is subject to the license terms in the LICENSE file found in the same directory.
+#
+# Copyright (C) 2021, Shenzhen Institute of Artificial Intelligence and Robotics for Society, all rights reserved.
+# Third party copyrights are property of their respective owners.
```
Comment (zihaomu) [inline models/text_detection_db/validation.py:5]: Please remove old copyrights.

File: models/text_detection_db/validation.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        with open(text_file, 'w+') as fid:
+            fid.write(result)
+    end = time.time()
+    avg_time=(end-start)/len(files)
+    result_dict = cal_recall_precison_f1(args.gt_dir,args.out_dir,default_evaluation_params)
+    # Print results
+    result_dict['avg time']=avg_time
+    print(result_dict)
+
+
```
Comment (zihaomu) [inline models/text_detection_db/validation.py:106]: Please keep a blank line at the end of the file.

File: models/text_detection_db/cal_rescall/rrc_evaluation_funcs.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    edge = [
+                ( point[1][0] - point[0][0])*( point[1][1] + point[0][1]),
+                ( point[2][0] - point[1][0])*( point[2][1] + point[1][1]),
+                ( point[3][0] - point[2][0])*( point[3][1] + point[2][1]),
+                ( point[0][0] - point[3][0])*( point[0][1] + point[3][1])
+    ]
+    
+    summatory = edge[0] + edge[1] + edge[2] + edge[3];
+    if summatory>0:
+        raise Exception("Points are not clockwise. The coordinates of bounding quadrilaterals have to be given in clockwise order. Regarding the correct interpretation of 'clockwise' remember that the train_images coordinate system used is the standard one, with the train_images origin at the upper left, the X axis extending to the right and Y axis extending downwards.")
```
Comment (zihaomu) [inline models/text_detection_db/cal_rescall/rrc_evaluation_funcs.py:278]: Please use newline for `Exception info`.

File: models/text_detection_db/cal_rescall/script.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+        sampleAP = 0
+
+        evaluationLog = ""
+
+        pointsList, _, transcriptionsList = rrc_evaluation_funcs.get_tl_line_values_from_file_contents(gtFile,
+                                                                                                       evaluationParams[
+                                                                                                           'CRLF'],
+                                                                                                       evaluationParams[
+                                                                                                           'LTRB'],
+                                                                                                       True, False)
```
Comment (zihaomu) [inline models/text_detection_db/cal_rescall/script.py:140]: wired format. Please use `\` for very long py codes.

File: models/text_detection_db/cal_rescall/rrc_evaluation_funcs.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+            raise Exception("Confidence value must be a float")       
+            
+    if withTranscription:
+        posTranscription = numPoints + (2 if withConfidence else 1)
+        transcription = m.group(posTranscription)
+        m2 = re.match(r'^\s*\"(.*)\"\s*$',transcription)
+        if m2 != None : #Transcription with double quotes, we extract the value and replace escaped characters
+            transcription = m2.group(1).replace("\\\\", "\\").replace("\\\"", "\"")
+    
+    return points,confidence,transcription
```
Comment (zihaomu) [inline models/text_detection_db/cal_rescall/rrc_evaluation_funcs.py:246]: A space needs to be added after each `,`.

**Review by zihaomu — COMMENTED** [review]
Thanks for your contribution.
There are some code style issues. Please check the [OpenCV code style of Python](https://github.com/opencv/opencv/wiki/Coding_Style_Guide#python-samples).

Is there a reference to `rrc_evaluation_funcs.py` file? If you refer to the code, please attach the original license. If you refer to the paper, please attach a link to the header of the file.

**Comment by zihaomu** [comment]
@the-star-sea, please refactor the code according to the test framework for other models. And after that, I think we can merge it.

**Comment by fengyuentau** [comment]
> 

Please, put the accuracy test in [tools/eval](https://github.com/opencv/opencv_zoo/tree/master/tools/eval). Benchmark is only for speed.

**Comment by the-star-sea** [comment]
> > 
> 
> Please, put the accuracy test in [tools/eval](https://github.com/opencv/opencv_zoo/tree/master/tools/eval). Benchmark is only for speed.

Only two days...I have just known this because my opencv_zoo is one month ago's type.

**Comment by fengyuentau** [comment]
> We dont put evaluation script along with models. Please take a look at https://github.com/opencv/opencv_zoo/tree/master/tools/eval.

This comment https://github.com/opencv/opencv_zoo/pull/73#issuecomment-1183923719 is from July 14.

**Comment by the-star-sea** [comment]
I forgot it😭

Yuantao Feng ***@***.***> 于 2022年9月8日周四 23:57写道：

**Comment by the-star-sea** [comment]
I still have enough time to add it I think.How can I update and get the
evaluation in my branch?

Stone Zhang ***@***.***> 于 2022年9月9日周五 00:48写道：

**Comment by fengyuentau** [comment]
```shell
# fetch upstream changes
git remote add upstream https://github.com/opencv/opencv_zoo.git
git fetch upstream
# store your local changes
git reset HEAD~5 #(5 is the number of your commits)
git stash
# merge upstream changes
git merge upstream/master
# restore your changes
git stash apply
# check and resolve conflicts
git status
```

**Comment by fengyuentau** [comment]
@the-star-sea friendly reminder.

**Comment by the-star-sea** [comment]
I will complete this after my graduate application period .I have discussed
it with Zihao before.

Yuantao Feng ***@***.***>于2022年10月29日 周六20:46写道：

**Comment by fengyuentau** [comment]
@the-star-sea Do you plan to continue working on this pull request in the near future?

**Comment by zihaomu** [comment]
Hi @the-star-sea, any update?

**Comment by fengyuentau** [comment]
Closing for no activities.

**Pull request closed without merging**